File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # my_set = set()
        
        # for i in range(0,len(nums) - 2) :
        #     for j in range(i+1,len(nums) - 1) :
        #         for k in range(j + 1, len(nums)):
        #             if nums[i] + nums[j] + nums[k] == 0 :
        #                 li = [nums[i] , nums[j] , nums[k]]
        #                 li[:] = sorted(li)
        #                 my_set.add(tuple(li))
        # return list(my_set)
        
        ans = []
        nums[:] = sorted(nums)
        
        for i in range(0,len(nums)) :
            if i > 0 and nums[i] == nums[i-1] :
                continue
            j , k = i + 1 , len(nums) - 1
            while j < k :
                sum_ = nums[i] + nums[j] + nums[k]
                if sum_ < 0 :
                    j += 1
                elif sum_ > 0 :
                    k -= 1
                else :
                    ans.append([nums[i],nums[j],nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j-1] == nums[j] : 
                        j += 1
                    while j < k and nums[k+1] == nums[k] : 
                        k -= 1
            
        return ans
          

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
//         Set<List<Integer>> mySet = new HashSet<>();
//         Arrays.sort(nums);
//         for(int i=0;i<nums.length;i++){
//             for(int j=i+1;j<nums.length;j++){
//                 for(int k=j+1;k<nums.length;k++){
//                     int sum = nums[i] + nums[j] + nums[k];
//                     if(sum == 0){
//                         List<Integer> triplet = new ArrayList<>();
//                         triplet.add(nums[i]);
//                         triplet.add(nums[j]);
//                         triplet.add(nums[k]);
//                         mySet.add(triplet);
//                     }
//                 }
//             }
//         }
        
//         List<List<Integer>> ans = new ArrayList<>();
        
//         for(List<Integer> triplet : mySet){
//             ans.add(triplet);
//         }
        
//         return ans;
        
        
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        
        for(int i=0;i<nums.length-2;i++){
            if(i == 0 || nums[i] != nums[i-1]){
                int low = i+1;
                int high = nums.length - 1;
                int sum = -nums[i];
                while(low < high){
                    if(nums[low] + nums[high] == sum){
//                         List<Integer> triplet = new ArrayList<>();
//                         triplet.add(nums[i]);
//                         triplet.add(nums[low]);
//                         triplet.add(nums[high]);
                        
//                         ans.add(triplet);
                        ans.add(Arrays.asList(nums[i], nums[low], nums[high]));
                        
                        while(low < high && nums[low] == nums[low+1]) low++;
                        while(low < high && nums[high] == nums[high-1]) high--;
                        
                        low++;
                        high--;
                    }
                    else if(nums[low] + nums[high] < sum) low++;
                    else high--;
                }
            }
        }
        
        return ans;
    }
}